cmake/ccfeatures.cmake

```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2019 Facebook.
+# SPDX-License-Identifier: Apache-2.0
+
+# Check for features like compiler builtins and generate the <ccfeatures.h> header file.
+include(CheckCSourceCompiles)
+
+if (PROJECT_BINARY_DIR)
+  set(CCFEATURES_H ${PROJECT_BINARY_DIR}/include/generated/ccfeatures.h)
+else ()
+  set(CCFEATURES_H ${CMAKE_BINARY_DIR}/include/generated/ccfeatures.h)
+endif ()
+
+check_c_source_compiles("
+  int main(void) {
+    int x=0; return __builtin_add_overflow(2, 3, &x);
+  }
+" HAVE___BUILTIN_ADD_OVERFLOW)
+
+check_c_source_compiles("
+  int main(void) {
+    int x=0; return __builtin_mul_overflow(2, 3, &x);
+  }
+" HAVE___BUILTIN_MUL_OVERFLOW)
+
+check_c_source_compiles("
+  int main(void) {
+    return __builtin_clz(2);
+  }
+" HAVE___BUILTIN_CLZ)
+
+check_c_source_compiles("
+  int main(void) {
+    return __builtin_clzll(2);
+  }
+" HAVE___BUILTIN_CLZLL)
+
+check_c_source_compiles("
+  int main(void) {
+    return __builtin_ctz(2);
+  }
+" HAVE___BUILTIN_CTZ)
+
+check_c_source_compiles("
```

Comment:
This must be integrated with the ZEPHYR_TOOLCHAIN_CAPABILITY_CACHE to reach an acceptable cmake configure performance.